File: src/main/scala/algorithms/union_find/java/PercolationStats.java
package algorithms.union_find.java;

import edu.princeton.cs.algs4.StdRandom;
import edu.princeton.cs.algs4.StdStats;

public class PercolationStats {

    private static final double CONFIDENCE = 1.29;
    private static final double ZERO = 0.0;
    private final double[] thresholds;
    private final int n;
    private final int trials;
    private double mean;
    private double stddev;


    public PercolationStats(int n, int trials) {
        if (n <= 0 || trials <= 0)
            throw new IllegalArgumentException();
        this.n = n;
        this.trials = trials;
        thresholds = new double[trials];
        for (int i = 0; i < thresholds.length; i++)
            thresholds[i] = this.runSimulation(new Percolation(n));
    }

    private double runSimulation(Percolation percolation) {
        while (!percolation.percolates()) {
            percolation.open(StdRandom.uniform(1, n+1), StdRandom.uniform(1, n+1));
        }
        return percolation.numberOfOpenSites() / (double) (n * n);
    }


    // sample mean of percolation threshold
    public double mean() {
        if (mean == ZERO)
            mean = StdStats.mean(thresholds);
        return mean;
    }

    // sample standard deviation of percolation threshold
    public double stddev() {
        if (stddev == ZERO)
            stddev = StdStats.stddev(thresholds);
        return stddev;
    }

    // low endpoint of 95% confidence interval
    public double confidenceLo() {
        return this.mean() - ((CONFIDENCE * this.stddev()) / Math.sqrt(trials));
    }

    // high endpoint of 95% confidence interval
    public double confidenceHi() {
        return this.mean() + ((CONFIDENCE * this.stddev()) / Math.sqrt(trials));
    }

    // test client (see below)
    public static void main(String[] args) {
        int n = Integer.parseInt(args[0]);
        int trials = Integer.parseInt(args[1]);
        PercolationStats percolationStats = new PercolationStats(n, trials);
        System.out.println("mean                    = " + percolationStats.mean());
        System.out.println("stddev                  = " + percolationStats.stddev());
        System.out.println("95% confidence interval = [" + percolationStats.confidenceLo() + ", " + percolationStats.confidenceHi() + "]");
    }
}

File: src/main/scala/algorithms/union_find/PercolationStats.scala
package algorithms.union_find

import scala.collection.parallel.ParSeq
import scala.util.Random

//monte carlo simulation
class PercolationStats(val n: Int, val trial: Int) {

  val simulations: Seq[Percolation] = Seq.tabulate(trial)(_ => new Percolation(n))
  val thresholds: ParSeq[Double] = simulations.par.map(runSimulation(_))

  private def runSimulation(percolation: Percolation): Double = {
    while(!percolation.percolates)
      percolation.open(Random.nextInt(n), Random.nextInt(n))
    percolation.numberOfOpenSites / (n * n).toDouble
  }

  // sample mean of percolation threshold
  def mean: Double = thresholds.sum / thresholds.size

  // sample standard deviation of percolation threshold
  def stddev: Double = ???

  // low endpoint of 95% confidence interval
  def confidenceLo: Double = ???

  // high endpoint of 95% confidence interval
  def confidenceHi: Double = ???
}


object PercolationStatsApp extends App {
 val percolationStats = new PercolationStats(200, 100)
  println(percolationStats.mean)

}
